dashboard/modules/job/job_head.py

```diff
@@ -204,6 +212,80 @@ async def list_jobs(self, req: Request) -> Response:
             content_type="application/json",
         )
 
+    @routes.get("/api/all_jobs/")
+    @optional_utils.init_ray_and_catch_exceptions(connect_to_serve=False)
+    async def list_jobs(self, req: Request) -> Response:
+        driver_jobs, submission_job_drivers = await self._get_driver_jobs()
+        submission_jobs = self._job_manager.list_jobs()
+        for job_id, job in submission_jobs.items():
+            job.driver = submission_job_drivers.get(job_id)
+        return Response(
+            text=json.dumps(
+                [
+                    *[
+                        {
+                            **dataclasses.asdict(job_info),
+                            "id": job_id,
+                            "type": "submission",
+                        }
+                        for job_id, job_info in submission_jobs.items()
+                    ],
+                    *[
+                        {**dataclasses.asdict(job_info), "id": job_id, "type": "driver"}
+                        for job_id, job_info in driver_jobs.items()
+                    ],
+                ]
+            ),
+            content_type="application/json",
+        )
+
+    async def _get_driver_jobs(self):
+        """
+        Returns a tuple of dictionaries related to drivers.
+
+        The first dictionary contains all driver jobs and is keyed by the job's id.
+        The second dictionary contains drivers that belong to submission jobs.
+        It's keyed by the submission job's id. Only the last driver of a submission job is returned.
+        """
+        request = gcs_service_pb2.GetAllJobInfoRequest()
+        reply = await self._gcs_job_info_stub.GetAllJobInfo(request, timeout=5)
+
+        jobs = {}
+        submission_job_drivers = {}
+        for job_table_entry in reply.job_info_list:
+            job_id = job_table_entry.job_id.hex()
+            metadata = dict(job_table_entry.config.metadata)
+            job_submission_id = metadata.get(JOB_ID_METADATA_KEY)
+            if not job_submission_id:
+                driver = DriverInfo(
+                    id=job_id,
+                    ip_address=job_table_entry.driver_ip_address,
+                    pid=job_table_entry.driver_pid,
+                )
+                job = JobInfo(
+                    status=JobStatus.SUCCEEDED
+                    if job_table_entry.is_dead
+                    else JobStatus.RUNNING,
+                    entrypoint="",
```

Comment:
I guess now we're using JobInfo to describe both "drivers" and "submission jobs".  Do we need to redefine some of the fields to adapt it to both?  For example `entrypoint` wasn't optional, because all submission jobs had entrypoints, but now perhaps it should be optional because drivers don't have them.

This would break backward compatibility though for consumers of the snapshot API, so maybe we shouldn't do this.

Reply:
I think we will eventually make this change when we move over to pydantic.